### PR TITLE
Update ci and ABI xml files to validate function parameters

### LINKED ISSUE #22712: Enchance libabigail ci check to validate function args/parameter/etc
currently, libabigail in our ci runs only validates the exported symbols that openssl produces.  We should also be tracking parameters to those functions.  To do this we need to enable dwarf information in the build that ci operates on for a proper comparison

### PR DESCRIPTION
**Pull request by nhorman**
libabigail is currenly only validating symbol presence and version information in ci.  We should also be validating function parameters, types, etc.  To do this we need to build the library with -g so the dwarf information is available for libabigail to interrogate

while we're at it, also add a script to re-generate the xml that abidiff uses for comparison during ci runs, to make updates easier

Fixes #22712

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -279,7 +279,7 @@ jobs:
     - name: install extra config support
       run: sudo apt-get -y install libsctp-dev abigail-tools libzstd-dev zstd
     - name: config
-      run: ./config --banner=Configured --strict-warnings enable-ktls enable-fips enable-egd enable-ec_nistp_64_gcc_128 enable-md2 enable-rc5 enable-sctp enable-ssl3 enable-ssl3-method enable-trace enable-zlib enable-zstd && perl configdata.pm --dump
+      run: CFLAGS="-g" ./config --banner=Configured --strict-warnings enable-ktls enable-fips enable-egd enable-ec_nistp_64_gcc_128 enable-md2 enable-rc5 enable-sctp enable-ssl3 enable-ssl3-method enable-trace enable-zlib enable-zstd && perl configdata.pm --dump
```
Comment (t8m) [inline .github/workflows/ci.yml:282]: Just put -g among the `./config` command line options.

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
The only thing that worries me is that the ABI .xml files are now really huge and they are not reviewable in the github web interface. We'll see how this will work on PRs that add new API calls.

**Comment by hlandau** [comment]
> The only thing that worries me is that the ABI .xml files are now really huge and they are not reviewable in the github web interface. We'll see how this will work on PRs that add new API calls.

This is a fair point. @nhorman I wonder if we could break these files down into smaller pieces somehow, maybe have a separate file for symbols beginning with a, b, c, etc.?

**Comment by nhorman** [comment]
possibly,I can look into it.  The abigail developers are also interested in any features we might like, so we could also ask them if some level of sharding isn't currently available

**Comment by nhorman** [comment]
Just to note it here, there is also an issue with updating the xml file, in that incorporating changes requires that it be built in exactly the same environment that the reference xml was built in - i.e. if I rebuild it on my system, it changes from what ci computes because the fedora 39 compiler enables intel CET, whereas the ubuntu ci image compiler does not, and we have ifdefs that change exposed structure members.  Its not a huge deal, but requires running the update in an ubuntu 22.04 container.

**Comment by nhorman** [comment]
It might also be worth noting, that in the nominal case, the abi file shouldn't change often.  There will be large changes on major version changes, as we purposely change/add/remove function signatures, but on released branches, and for the majority of the development process.  If the file changes most of the time, the xml shouldn't change as a change represents an abi break

**Comment by t8m** [comment]
Any API addition will break the CI. And these are common on the master branch. But unless the diff is too huge outputting the diff as part of the CI as is done in #22689 should help.

**Comment by nhorman** [comment]
@t8m, yes, exactly.  While it may be common to add apis, the delta when the xml file is computed will be small (i.e. some small number of lines per api symbol changed/added), and should be manageable for review.

times that I can think of when the file will not be small are when we update library versions (i.e. moving from OPENSSL_3_0_0 to OPENSSL_3_0_1 will alter every symbol and the abi xml will effectively be a completely new file.  In these cases I'm not sure what we can do about them (in fact splitting the file into some consumable chunk may actually make the problem worse, as it means completely changing X files rather than a single file)

One apprach that may be feasible is not storing the xml at all, and instead doing a separate build of a commit prior to the first commit of the PR.  Then you can just compare binary DSO's easily enough.  That entails a full extra build cycle, which I'm not wild about.

**Comment by hlandau** [comment]
What are we doing to determine our "fips changed" label right now? Might be worth looking at.

**Comment by nhorman** [comment]
looks like we maintain a list of fips-relevant files, and for each file in that list, we use a C compressor to represent that file as a string, on which we compute the sha256 sum.  differences in the current files sum to a recorded previous file trigger the fips_changed label.

Thats probably a bit heavy handed for what we want to do with abi validation.  That will trigger change reviews on changes to quoted strings, etc, which we don't want here, and its still going to produce a large output over the whole source tree, as opposed to fips  (641 files for fips vs 2172 c/h/S/s files for the whole tree)

**Comment by t8m** [comment]
Merged to the master branch. Thank you.

**Pull request closed without merging**